README.md
# Ethereum-Hot-Wallet
* Enviorment
    *  python Go version: 1.9.2
    * Python library: Flask==0.12.1, web3==3.16.4
    * Go version: 1.9.2
    * geth version: v1.8.0-unstable-56152b31
    
* geth command: geth -rpc -rpcaddr "localhost" --rpcport 8545  --rpcapi web3,eth,miner,admin --datadir "./chain" --nodiscover console 2>>eth_output.log 
---
* API:
    - /admin.nodeinfo 
    - /miner method: PUT start miner .
    - /miner method: DELETE stop miner .
    - /block method: GET query block by number.
        - request:
            - number (Integer)
        - response:
            - result
                - {
                    - difficulty  (Integer)
                    - gasLimit (Integer)
                    - gasUsed (Integer)
                    - hash (String)
                    - miner (String)
                    - parentHash (String)
                    - totalDifficulty  (Integer)
                - }
    - /transaction method: GET query transaction by hash.
        - request:
            - hask (String)
        - response:
            - result
                - {
                    - blockHash  (String)
                    - blockNumber (Integer)
                    - from (String)
                    - gas (Integer)
                    - gasPrice (Integer)
                    - hash (String)
                    - nonce (Integer)
                    - to (String) 
                    - value  (Integer)
                - }
    
* Error response format
    - {
        - error: {
              message : "Error message"
        - }
    - }

* Basic http error code
    - 400 for wrong parameter format.
        - {
            - error: {
                  message : "Error message"
            - }
        - }
    - 404 for not found of block, transcation.
        - {
            - error: {
                  message : "Error message"
            - }
            - result : null
        - }
    - 500 for other error which is responsibility of this api server, log.




chain/history
> web3.eth.accounts
web3.eth.accounts
miner.start(1)
miner.stop()
web3.eth.getBalance('0xe73c6aa49d824a83f57f7558110560bb042e57fc')
web3.eth.accounts
acc0 = web3.eth.accounts[0]
acc1 = web3.eth.accounts[1]
acc0
acc1
web3.eth.sendTransaction({from:acc0,to:acc1,value:web3.toWei(3,"ether")})
web3.eth.sendTransaction({from:acc0,to:acc1,value:web3.toWei(1,"ether")})
web3.eth.sendTransaction({from:acc0,to:acc1,value:web3.toWei(0,"ether")})
miner.start(1)
web3.eth.sendTransaction({from:acc0,to:acc1,value:web3.toWei(1,"ether")})
function checkAllBalances() {
     var totalBal = 0;
     for (var acctNum in eth.accounts) {
         var acct = eth.accounts[acctNum];
         var acctBal = web3.fromWei(eth.getBalance(acct), "ether");
         totalBal += parseFloat(acctBal);
         console.log("  eth.accounts[" + acctNum + "]: \t" + acct + " \tbalance: " + acctBal + " ether");
     }
     console.log("  Total balance: " + totalBal + " ether");
 };
checkAllBalances();
web3.eth.sendTransaction({from:acc0,to:acc1,value:web3.toWei(1,"ether")})
checkAllBalances();
miner.stop()
admin.nodeInfo()
checkAllBalances();
miner.stop()
admin.nodeInfo
eth.getBlock
eth.getBlock()
eth.getBlock('0x5e1fc79cb4ffa4739177b5408045cd5d51c6cf766133f23f7cd72ee1f8d790e0')
eth.getBlock('0x5e1fc79cb4ffa4739177b5408045cd5d51c6cf766')
eth.getBlock()
eth.getBlockNumber()
eth.getBlockNumber
eth.getAccounts()
eth.getAccounts
eth.getBlock(1000)
eth.getBlock(0)
eth.getBlock(1)
eth.getBlock(2)
eth.eth.getTransactionCount
eth.eth.getTransactionCount()
eth.eth.getTransactionCount(1)
eth.eth.getTransactionCount(0)
checkAllBalances();
eth
eth.getBlock(100)
eth.getBlock(2)
eth.getTransactionFromBlock(0)
eth.getTransactionFromBlock("0x57e545b4e0c506ec412e59319983c73b1a557ed46509eb0f4a9034504cd7a56a")
eth.getTransactionCount()
eth.getTransactionCount
eth.getAccounts
eth.getAccounts()
miner.start()
miner.stop()
eth.getAccounts()
eth.getAccounts
eth.getAccounts()
eth.accounts
miner.stop()
miner.start(4)
miner.start
miner.start()

api.py
# -*- coding: utf-8 -*-
from flask import Flask, jsonify, abort, request
from contextlib import contextmanager
from requests.exceptions import ConnectionError, ConnectTimeout
from web3 import Web3, HTTPProvider
import json
import requests
import logging
from helpers import convert_filter_keys

app = Flask(__name__)
app.config['geth_url'] =  'http://127.0.0.1:8545'

def logging(message):
	print("log: %s" % message)


@contextmanager
def context_handler(*exceptions):
  try:
    yield
  except ValueError as valueError:
  	logging(type(valueError))
  	abort(400, {'message': 'parameter hash value is invalid.'});
  except ConnectTimeout as connectionTimeout:
  	logging("ConnectionTimeout %s" % app.config['geth_url'])
  	abort(500, {'message': 'Internal Server Error'});
  except ConnectionError as connectionError:
  	logging("ConnectionError %s" % app.config['geth_url'])
  	abort(500, {'message': 'Internal Server Error'});
  else:
  	pass


@app.route("/admin.nodeinfo", methods=["GET"])
def admin_nodeinfo():
	w3 = Web3(HTTPProvider(app.config['geth_url']))
	return jsonify({'message': 'ok', 'result': convert_filter_keys(w3.admin.nodeInfo, ['enode', 'name'])})


@app.route("/block")
def block():
	# set up connection of geth client.
	w3 = Web3(HTTPProvider(app.config['geth_url']))
	with context_handler(Exception):
		# query: getBlock by block number.
		# get block number parameter from GET request.
		block_number = int(request.args.get('number', -1))
		block = w3.eth.getBlock(block_number)
		# Null Traction object.
		if block is None:
			abort(404, {'message': 'Block number %i not found' % block_number})

	return jsonify({'message': 'ok', 'result': convert_filter_keys(block, ['difficulty', 'gasLimit', 'gasUsed', 'hash', 'miner', 'parentHash', 'totalDifficulty'])})

@app.route("/transaction", methods=['GET'])
def transaction():
	w3 = Web3(HTTPProvider(app.config['geth_url']))
	with context_handler(Exception):
		hash = str(request.args.get('hash', ''))
		trasaction = w3.eth.getTransaction(hash)
		if trasaction is None:
			abort(404, {'message': 'Transaction hash %s not found' % hash})
	
	return jsonify({'message': 'ok', 'result': convert_filter_keys(trasaction, ['blockHash', 'blockNumber', 'from', 'gas', 'gasPrice', 'hash', 'nonce', 'to', 'value'])})

@app.route("/miner", methods=["PUT", "DELETE"])
def miner():
	w3 = Web3(HTTPProvider(app.config['geth_url']))
	if request.method == "PUT":
		numberOfThread = 4
		w3.miner.start(numberOfThread)
		return jsonify({'message': 'miner start'})
	elif request.method == "DELETE":
		w3.miner.stop()
		return jsonify({'message': 'miner stop'})

	return jsonify({'message': 'not thnig'})

@app.errorhandler(400)
def custom400(error):
	return jsonify({'error': {'message': error.description['message']}}), 400

@app.errorhandler(404)
def custom404(error):
	print('404 handler')
	return jsonify({'error': {'message': error.description['message']}, 'result': None}), 404

@app.errorhandler(500)
def internal_error(error):
	return jsonify({'error': {'message': error.description['message']}}), 500

if __name__ == '__main__':
	app.run(debug=True, host='0.0.0.0', port=5678)